Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp

```diff
@@ -1153,13 +1204,70 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addCatchAll(Stack&, ControlType
 
 PartialResult WARN_UNUSED_RETURN IPIntGenerator::addCatchAllToUnreachable(ControlType& block)
 {
-    block = ControlType(block.signature(), BlockType::Catch, CatchKind::CatchAll);
+    if (ControlType::isTry(block))
+        convertTryToCatch(block, CatchKind::CatchAll);
+    else
+        block.m_catchKind = CatchKind::CatchAll;
+
+    // FIXME: If this is actually unreachable we shouldn't need metadata.
+    auto size = m_metadata->m_metadata.size();
+    block.m_awaitingUpdate.append(size);
+    m_metadata->addBlankSpace(8);
+
+    m_metadata->m_exceptionHandlers.append({
+        HandlerType::CatchAll,
+        static_cast<uint32_t>(block.m_pc),
+        static_cast<uint32_t>(block.m_pcEnd),
+        static_cast<uint32_t>(m_parser->offset() - m_metadata->m_bytecodeOffset),
+        static_cast<uint32_t>(m_metadata->m_metadata.size()),
+        m_tryDepth,
+        0
+    });
+
+    m_metadata->addBlankSpace(4);
+    // IPInt stack entries are 16 bytes to keep the stack aligned. With the exception of locals, which are only 8 bytes.
+    uint32_t stackSizeInV128 = m_parser->getStackHeightInValues() + roundUpToMultipleOf<2>(m_metadata->m_numLocals) / 2;
+    WRITE_TO_METADATA(m_metadata->m_metadata.data() + size + 8, stackSizeInV128, uint32_t);
+
+    return { };
+}
+
+PartialResult WARN_UNUSED_RETURN IPIntGenerator::addDelegate(ControlType& target, ControlType& data)
+{
+    return addDelegateToUnreachable(target, data);
+}
+PartialResult WARN_UNUSED_RETURN IPIntGenerator::addDelegateToUnreachable(ControlType& target, ControlType& data)
+{
+    // FIXME: If this is actually unreachable we shouldn't need metadata.
+    auto size = m_metadata->m_metadata.size();
+    data.m_awaitingUpdate.append(size);
+    m_metadata->addBlankSpace(8);
+
+    ASSERT(ControlType::isTry(target) || ControlType::isTopLevel(target));
+    unsigned targetDepth = ControlType::isTry(target) ? target.m_tryDepth : 0;
+
+    m_metadata->m_exceptionHandlers.append({
+        HandlerType::Delegate,
+        static_cast<uint32_t>(data.m_pc),
+        static_cast<uint32_t>(data.m_pcEnd),
```

Comment:
What does this do if we delegate to a catch-less try? It looks to me like we only ever set endPC in the catch, right?